BRANCHING.md

```diff
@@ -0,0 +1,167 @@
+
+* After making a change in the release branch
+* On a periodic basis
+* After finishing the release
+
+Currently, we want this merge to happen via a PR and go through (usually perfunctory)
+review. Reviewers should check that this is not a squash merge or a rebase merge.
+
+Right now we think it's desirable that whenever the release branch moves forwards,
+a PR could be opened automatically that proposes to merge the release branch to master,
+exposing any git conflicts, which can then be reviewed and resolved in this PR.
+
+Making changes
+--------------
+
+Developers make changes using PRs the same way as always, and are free to
+squash-merge their development PRs as they historically have. (It doesn't much
+matter for this workflow whether or not developer PRs are squash merges.)
+
+When a developer wants to make a change, they should ask themselves, should this
+go in the latest release branch, or in master?
+
+* Changes which go to the release branch will eventually be merged forwards to
+  master. No cherry-picking will be necessary. This is usually the right thing
+  for a bug fix.
+* Changes which should NOT go into the release branch should just go to master.
+
+This decision tree covers the majority of changes, and does not involve any
+cherry-picking.
+
+Sometimes, a developer will make a PR to master, but then later realize it
+should have gone to the release branch.
+
+* At this point, you have to cherry-pick the change back to the release branch.
+* You can cherry-pick your commit(s) from master.
+* You should expect a git conflict to occur when release is merged into master
+  after this. This is because a cherry-picked commit has a different hash from
+  the original commit, and git does not consider them "the same" when merging,
+  they are instead very similar commits touching the same code.
+* You will have to resolve this conflict, which usually won't be too hard.
+* To avoid this situation, make sure you consider whether a change should go
+  to master or the release branch. If in doubt, ask in the channel.
+
+This leads to a git history like the following:
+
+[![](https://mermaid.ink/img/pako:eNqtU8FuwjAM_ZXIZ7aBaJbR86addto1F5OYNippq5CiIcS_L5QxlS2FViKnOO8lfn6O96AqTZACCysz_t1hncuStUtV1hofj5YOS5UzR2vCDT1tk0saMzqVgAlfSIghy2fx8g_xmAVomzxOO1BOqqgazyyaMi7FksuoV8mg6JzkejlqNu0pRwsxv6F5oMqbSn5Nmt0n4Yge81jpxPk8bspKiGRUI7uXvxYdq_-Uwvvd6kLO7R5qo4rIg4M-wxUzBym46Bcf-alPEUwgZA4sHUZ0f8Qk-JwsSUjDVqMrjq8eAg8bX33uSgWpdw1NoKk1eno1mDm0kK5wvQmnpI2v3Mdp5tvRPzPfWuSHePgGwSE-5w?type=png)](https://mermaid.live/edit#pako:eNqtU8FuwjAM_ZXIZ7aBaJbR86addto1F5OYNippq5CiIcS_L5QxlS2FViKnOO8lfn6O96AqTZACCysz_t1hncuStUtV1hofj5YOS5UzR2vCDT1tk0saMzqVgAlfSIghy2fx8g_xmAVomzxOO1BOqqgazyyaMi7FksuoV8mg6JzkejlqNu0pRwsxv6F5oMqbSn5Nmt0n4Yge81jpxPk8bspKiGRUI7uXvxYdq_-Uwvvd6kLO7R5qo4rIg4M-wxUzBym46Bcf-alPEUwgZA4sHUZ0f8Qk-JwsSUjDVqMrjq8eAg8bX33uSgWpdw1NoKk1eno1mDm0kK5wvQmnpI2v3Mdp5tvRPzPfWuSHePgGwSE-5w)
```

Comment:
:question: curious why an image link versus the raw mermaid text?  I imagine if mermaid.live goes down mermaid rendering may be gone as well so probably no harm linking to the image.

Reply:
IDK, I wanted to do inline mermaid, but it seems like the github markdown doesn't support the gitGraph mermaid type. The mermaid.live editor gave me this blob to paste right into markdown so that's what I went with.